### ISSUE TITLE
SMTP - SSL shared cipher

### ISSUE THREAD

**Issue by seannaswell** [zotonic/zotonic#3686]
## `Zotonic` version
Latest master

I have several websites relaying email thru an email server, and it has worked great for several months. A few days ago I noticed that mail is no longer being sent, so I checked the logs in Zotonic and the mail server.

The mail server reports the same error each time Zotonic attempts to relay a message: `error:0A0000C1:SSL routines:no shared cipher`

Nothing has changed on the mail server, and have verified everything is good using the [internet.nl](https://internet.nl/test-mail/) test site, so I am wondering if something may have recently changed in Zotonic related to SMTP? 

I can setup a test VM and do the same test to a Zotonic instance setup to receive mail, but just wanted to make an inquiry first, in case there was something changed recently that is quick and simple to diagnose...

**Comment by mworrell** [comment]
Nothing has changed on our end, at least not regarding ciphers (afaik).

Which Erlang version do you use?  Might be that the SMTP servers started using newer ciphers not yet in the Erlang version.

We should be falling back to port 25 unencrypted if the TLS connection could not be made.

**Comment by seannaswell** [comment]
The mail server uses OpenSSL v3.0.13 and Erlang version is 25.3.2.7.

I do notice that Zotonic retries and says it's attempting without TLS, but the mail server logs show TLS is still attempted. 

You answered my question, it's not something quick and easy. Will do a little more digging and get back with more details.

**Comment by alvaropag** [comment]
Hello,

Maybe the changes made in this issue?

https://github.com/zotonic/zotonic/issues/3481

--
Álvaro Pagliari

**Comment by mworrell** [comment]
Indeed, thinking of it, we did start passing extra TLS options. They are meant for the certificate checks.

**Comment by seannaswell** [comment]
The changes mentioned [(#3481)](https://github.com/zotonic/zotonic/issues/3481#issuecomment-1747506024) do appear to include a directive, `{tls, always},` which it seems would prevent falling back to non-TLS if there was an issue cipher library compatibility.

Maybe the newer version of OpenSSL invalidated some ciphers being used in Erlang 25? If Zotonic did fall back to non-TLS transmission when/if there is an issue finding compatible ciphers, then worst case in situations like this would be a delay sending mail.

Again, will follow up after more testing, just wanted to mention the `tls always` directive, as that's one thing I noticed looking at the logs...

**Comment by seannaswell** [comment]
Okay, a few things to report...

First, according to the [Erlang SSL manpage](https://www.erlang.org/doc/man/ssl.html), TLS versions 1.0 and 1.1 are now considered Legacy, and TLSv1.3 has been added.

However, according to the [gen_smtp](https://github.com/gen-smtp/gen_smtp) README, TLS versions 1.0 and 1.1 are still included by default, and TLSv1.3 is not listed. 

When making the following change, the `no shared cipher` message goes away:

`{tls_options, [
            {versions, ['tlsv1.3, tlsv1.2']}...`

However, the connection fails, likely because TLSv1.3 isn't actually supported.

So, it appears that OpenSSL 3.0.13 now requires TLS1.3, but gen_smtp doesn't support it.

Once gen_smtp is updated to remove TLS1.0 and TLS1.1, and TLS1.3 is added, the `tls_options version` command can likely be removed. At least that is what it looks like to me, but I could be completely wrong...

**Comment by seannaswell** [comment]
I have disabled TLS on the mail server for connections from Zotonic, as they are on the same LAN. That solves the immediate issue for me.

Unfortunately I do not understand why TLSv1.2 from Zotonic does not work, while TLS1.2 connections from other email servers does. Perhaps updating `gen_smtp` will solve that.

If there is any further testing I can do, please let me know.

**Comment by mworrell** [comment]
@seannaswell Could you try #3695 to see if that fixes your problem?

I am not sure if we should default to port 25 or port 587 for sending messages. Right now we use 587 when first trying with TLS and 25 if we try without TLS.

/cc @CyBeRoni

**Comment by CyBeRoni** [comment]
Technically 587 is for submission, i.e. an MUA submitting emails to be relayed onwards. 25 is for MTA-to-MTA traffic.

So I believe defaulting to 587 would be more correct, since Zotonic acts as an MUA here.

The protocol is entirely identical, but relay servers may choose to treat port 587 differently.

**Comment by seannaswell** [comment]
@mworrell I will test [#3695](https://github.com/zotonic/zotonic/pull/3695) and follow-up.

@CyBeRoni is correct, reverting to port 25 is probably not a good idea. If a TLS connection fails, it can be attempted on the same port non-TLS.

Most people will be specifying a Username/Password and using port 587 to Forward to a big mail provider, where TLS will likely always work. If TLS fails for some reason, and a non-TLS Auth attempt is made on 587, the mail provider will likely reject it with a clear error message. So, changing ports isn't necessary. 

However, some people will be able to receive mail without TLS, so attempting without TLS, on the same port, should be an option.

**Comment by mworrell** [comment]
@seannaswell could you check which error `gen_smtp_client:send_blocking` returns for the TLS send-try?

The code is in z_email_server, line 1093:

```erlang
send_blocking_smtp(MsgId, VERP, RecipientEmail, EncodedMail, SmtpOpts, Context) ->
    {relay, Relay} = proplists:lookup(relay, SmtpOpts),
    ?LOG_INFO(#{
        text => <<"Sending email">>,
        in => zotonic_core,
        recipient => RecipientEmail,
        message_id => MsgId,
        relay => Relay,
        port => proplists:get_value(port, SmtpOpts)
    }),
    case gen_smtp_client:send_blocking({VERP, [RecipientEmail], EncodedMail}, SmtpOpts) of
```

**Comment by mworrell** [comment]
Regarding the port - I am not sure. When Zotonic is running as a server (instead of dev system) we deliver directly email to Google et al. Shouldn't we use port 25 for this server-to-server traffic?  Or is port 587 also ok for this traffic.

**Comment by seannaswell** [comment]
> Regarding the port - I am not sure. When Zotonic is running as a server (instead of dev system) we deliver directly email to Google et al. Shouldn't we use port 25 for this server-to-server traffic? Or is port 587 also ok for this traffic.

@CyBeRoni alluded to the different modes in his comment, and it is a bit tricky because Zotonic acts as an MTA when in server mode, and more like an MUA when in Forward mode.

It seems defaulting to port 25 in all cases is the easiest and least intrusive way to handle. 

If a user enables forwarding and overrides the default port, then all communication (TLS or non-TLS) should take place over the user specified port.

**Comment by seannaswell** [comment]
> @seannaswell could you check which error `gen_smtp_client:send_blocking` returns for the TLS send-try?

```
2024-02-28 01:52:48 INFO  site=seannaswell  user_id=1 <0.9781.0> [z_email_server:send_blocking_smtp/6:1095] 
 text="Sending email" relay=10.0.1.14 recipient=test@test.domain port=25 message_id=l54ku0zztghnx5pjmmft in=zotonic_core
2024-02-28 01:52:55 ERROR  site=seannaswell  user_id=1 <0.9781.0> [z_email_server:spawned_email_sender_loop/13:945] 
 text="Error sending email" result=error relay=10.0.1.14 recipient=test@test.domain reason=send message={error,closed} in=zotonic_core host=10.0.1.14 failure_type=network_failure
```

Looking at Debug the INFO notice references `send_blocking_smtp` and generates an `{error,closed}` message in the Email Log.
